executor/common_linux.h

```diff
@@ -1357,6 +1357,117 @@ static long syz_emit_ethernet(volatile long a0, volatile long a1, volatile long
 }
 #endif
 
+#if SYZ_EXECUTOR || __NR__syz_io_uring_put_sqes_on_ring
+//#include <linux/fs.h> // for __kernel_rwf_t
+
+// from liburing (tools/io_uring/barrier.h)
+#ifndef LIBURING_BARRIER_H
+#define LIBURING_BARRIER_H
+
+#if defined(__x86_64) || defined(__i386__)
+#define read_barrier() __asm__ __volatile__("" :: \
+						: "memory")
+#define write_barrier() __asm__ __volatile__("" :: \
+						 : "memory")
+#else
+/*
+ * Add arch appropriate definitions. Be safe and use full barriers for
+ * archs we don't have support for.
+ */
+#define read_barrier() __sync_synchronize()
+#define write_barrier() __sync_synchronize()
+#endif
+
+#endif
+
+// from linux/io_uring.h
+struct io_sqring_offsets {
+	__u32 head;
+	__u32 tail;
+	__u32 ring_mask;
+	__u32 ring_entries;
+	__u32 flags;
+	__u32 dropped;
+	__u32 array;
+	__u32 resv1;
+	__u64 resv2;
+};
+
+/**
+ * sizeof(io_uring_sqe) (as defined in linux/io_uring.h) is not likely to change -- there
+ * are currently paddings in the structure for backward compatibility in future versions.
+ * struct io_uring_sqe is only used for ptr arithmetic and size computation, thus, simply
+ * use a placeholder.
+ * */
+#define SIZEOF_IO_URING_SQE 64
+struct io_uring_sqe {
+	char fill[SIZEOF_IO_URING_SQE];
+};
+
+static void syz_io_uring_put_sqes_on_ring(volatile long a0, volatile long a1, volatile long a2, volatile long a3)
+{
+	// syzlang: syz_io_uring_put_sqes_on_ring(sq_ring_ptr sq_ring_ptr, sq_off ptr[inout, io_sqring_offsets], sqes ptr[in, array[ptr[in, io_uring_sqe]]], size len[sqes])
+	// C:       syz_io_uring_put_sqes_on_ring(void* sq_ring_ptr,       io_sqring_offsets* sq_off,            io_uring_sqe*** sqes,               size_t len_sqes)
+
+	// Cast to original
+	/* Offsets in io_uring_setup() are obtained using offsetof(), thus, they are in bytes. 
+	 * To expose those offsets in ptr arithmetic, use char* for sq_ring_ptr */
+	char* sq_ring_ptr = (char*)a0; // obtained using mmap$IORING_OFF_SQ_RING()
+	io_sqring_offsets* sq_off = (io_sqring_offsets*)a1;
```

Comment:
With the current descriptions syzkaller will not pass actual offsets here, it will pass just 0s. We could make all of head/tail/mask/etc resources, but then it will be hard to generate even the simplest correct program.
I think we should query these offsets once during executor startup (b/c they are constant for a particular kernel version) and then just use them here, w/o accepting from the program.

Reply:
The offsets are now computed within the executor.

Here is the computation of offsets that are constant per kernel version except `io_sqring_offsets.array`: [func def](https://github.com/google/syzkaller/pull/1926/files#diff-def271e38e744a7dca0b6e9075acb4f2R1446-R1471), [usage](https://github.com/google/syzkaller/pull/1926/files#diff-def271e38e744a7dca0b6e9075acb4f2R1492)

The computation of `io_sqring_offsets.array` is slightly more tricky since the number of sq and cq entries are needed (the computation stack: [`io_uring_create()`](https://elixir.bootlin.com/linux/v5.8-rc3/source/fs/io_uring.c#L8182) <- [`io_allocate_scq_urings()`](https://elixir.bootlin.com/linux/v5.8-rc3/source/fs/io_uring.c#L8024) <- [`rings_size()`](https://elixir.bootlin.com/linux/v5.8-rc3/source/fs/io_uring.c#L7239))

To simulate that, [`put_io_uring_sq_array_offset()`](https://github.com/google/syzkaller/pull/1926/files#diff-def271e38e744a7dca0b6e9075acb4f2R1412-R1444) was introduced to set `io_sqring_offsets.array` (mainly mimicks 
 [`rings_size()`](https://elixir.bootlin.com/linux/v5.8-rc3/source/fs/io_uring.c#L7239)).

We can't use [some macros](https://github.com/google/syzkaller/pull/1926/files#diff-def271e38e744a7dca0b6e9075acb4f2R1362) since we can't include `linux/overflow.h`. Thus, the code does not compile.

Maybe we can define `io_sqring_offsets.array` as the only resource among the others in `io_sqring_offsets` and avoid computing it.